### ISSUE TITLE
Reinforcement Learning from Human Feedback (RLHF) examples: Direct Preference Optimization (DPO)

### ISSUE THREAD

**Issue by danilopeixoto** [ml-explore/mlx-examples#513]
Introduce one Reinforcement Learning from Human Feedback (RLHF) example, such as Direct Preference Optimization (DPO) method.

**Paper**

[Direct Preference Optimization: Your Language Model is Secretly a Reward Model](https://arxiv.org/abs/2305.18290)

**Notes**

[Direct Preference Optimization (DPO): A Simplified Explanation by João Lages](https://medium.com/@joaolages/direct-preference-optimization-dpo-622fc1f18707)
![](https://miro.medium.com/v2/resize:fit:1400/format:webp/1*AqKOT0pxzi5kOgiobb-Fvg.png)

**Implementation example**

[huggingface/trl: TRL - Transformer Reinforcement Learning](https://github.com/huggingface/trl)

**Possible MLX implementation**

Policy and reference log probabilities:

```python
def get_batched_logps(model, inputs, targets, pad_token_id=0):
    logits, _ = model(inputs)
    logits = logits.astype(mx.float32)

    loss_mask = targets != pad_token_id
    per_token_logps = mx.take_along_axis(nn.log_softmax(logits), targets[..., None], axis=2).squeeze(2)

    return tuple((per_token_logps * loss_mask).sum(-1).split(2))
```

Loss:

```python
def dpo_loss(model, beta, reference_chosen_logps, reference_rejected_logps, inputs, targets):
    chosen_logps, rejected_logps = get_batched_logps(model, inputs, targets)

    pi_logratios = chosen_logps - rejected_logps
    reference_logratios = reference_chosen_logps - reference_rejected_logps

    logratios = pi_logratios - reference_logratios
    losses = -nn.log_sigmoid(beta * logratios)

    chosen_rewards = beta * (chosen_logps - reference_chosen_logps)
    rejected_rewards = beta * (rejected_logps - reference_rejected_logps)
    reward_accuracies = (chosen_rewards > rejected_rewards).astype(mx.float32)
    reward_margins = chosen_rewards - rejected_rewards

    return (
        losses.mean(),
        chosen_rewards.mean(),
        rejected_rewards.mean(),
        reward_accuracies.mean(),
        reward_margins.mean(),
    )
```

**Comment by awni** [comment]
@danilopeixoto I've been thinking about having this in MLX LM recently. Any interest in sending a PR?

It might make to do it after we have a more manageable config (https://github.com/ml-explore/mlx-examples/pull/503) but that should be landed soon!

**Comment by awni** [comment]
To be more concrete, I'm envisioning you just set the loss in the config. e.g. `cross_entropy` or `dpo`

**Comment by N8python** [comment]
I'm very very excited for this! Don't have the technical expertise to implement the DPO directly but would love to help in other ways (config, code cleanup) if neccessary!

**Comment by lin72h** [comment]
That makes MLX really useful for production not just a research tool!

**Comment by kishoretvk** [comment]
+500 waiting for this

**Comment by developerlin** [comment]
Wait for this, when will the DPO training is supported?

**Comment by Goekdeniz-Guelmez** [comment]
#1233 #1210 #1209
